### ISSUE TITLE
Streaming from URL ending in / causes memory corruption

### ISSUE THREAD

**Issue by low-batt** [mpv-player/mpv#11626]
### Important Information

Provide following Information:
- mpv version:  master
- macOS Version:  13.3.1
- Source of the mpv binary or bundle:  My own debug build using [mpv-build](https://github.com/mpv-player/mpv-build) and Xcode 14.3
- If known which version of mpv introduced the problem:  unknown
- Possible screenshot or video of visual glitches: N/A

### Reproduction steps

In a debug version of mpv running under `lldb` on the Mac and configured to use `libgmalloc.dylib` stream this video:
https://www.tv2.no/video/nyhetene/satt-ut-av-oppvaskmaskin-triks/1810348/

Before the mpv window appears execution will stop with a `EXC_BAD_ACCESS`.

I found that URL in the closed issue #10886.
Possibly this explains the heap corruption reported in issue #11378.

I built a debug version of mpv using [mpv-build](https://github.com/mpv-player/mpv-build) by doing this before building:
```bash
printf "%s\n" -Dbuildtype=debug > mpv_options
```

My debug session is shown below, but you may not need to look into it. This statement is corrupting memory: https://github.com/mpv-player/mpv/blob/4fd0a39fc8aad3fb8d7bb8ed2b2918e1e0e7a53b/misc/charset_conv.c#L232

The method `mp_iconv_to_utf8` has a bunch of input checking at the start of the method, but it is not checking to see if the string to convert passed in `buf` is empty. This code sequence in `append_dir_subtitles`:
https://github.com/mpv-player/mpv/blob/4fd0a39fc8aad3fb8d7bb8ed2b2918e1e0e7a53b/player/external_files.c#L188-L189

Is passing an empty string to `mp_iconv_to_utf8` because `mp_basename` finds the / at the end of the URL and returns an empty string. At least I think this is what is happening.

I added this code to the start of `mp_iconv_to_utf8`:
```c
    if (!buf.len)
      return buf;
```
With that change I was able to run `mpv` under `lldb` with  `libgmalloc.dylib` and successfully stream the video.

Once I get this issue posted and can refer to an issue number I will post a pull request with that change.  I am a total newbie at debugging `mpv` so a critical review of my analysis is needed.

<details>
<summary><b>lldb session:</b></summary>

```bash
low-batt@gag mpv-build (master %=)$ lldb ./mpv/build/mpv -- --config=no  --log-file=mpv.log --mute https://www.tv2.no/video/nyhetene/satt-ut-av-oppvaskmaskin-triks/1810348/
(lldb) target create "./mpv/build/mpv"
Current executable set to '/Users/low-batt/Documents/builds/mpv-build/mpv-build/mpv/build/mpv' (arm64).
(lldb) settings set -- target.run-args  "--config=no" "--log-file=mpv.log" "--mute" "https://www.tv2.no/video/nyhetene/satt-ut-av-oppvaskmaskin-triks/1810348/"
(lldb) env DYLD_INSERT_LIBRARIES=/usr/lib/libgmalloc.dylib
(lldb) process launch
Process 84056 launched: '/Users/low-batt/Documents/builds/mpv-build/mpv-build/mpv/build/mpv' (arm64)
GuardMalloc[mpv-84056]: Allocations will be placed on 16 byte boundaries.
GuardMalloc[mpv-84056]:  - Some buffer overruns may not be noticed.
GuardMalloc[mpv-84056]:  - Applications using vector instructions (e.g., SSE) should work.
GuardMalloc[mpv-84056]: version 064559.72.1
2023-04-24 16:43:47.014580-0400 mpv[84056:2489292] [Entitlements] MSVEntitlementUtilities - Process mpv PID[84056] - Group: (null) - Entitlement: com.apple.mediaremote.external-artwork-validation - Entitled: NO - Error: (null)
Process 84056 stopped
* thread #18, name = 'mpv/worker', stop reason = EXC_BAD_ACCESS (code=1, address=0x2dc1d8000)
    frame #0: 0x00000001000ecdec mpv`mp_iconv_to_utf8(log=0x00000002dc1affc0, buf=(start = "", len = 0), cp="UTF-8-MAC", flags=8) at charset_conv.c:232:31
   229 	
   230 	    iconv_close(icdsc);
   231 	
-> 232 	    outbuf[osize - oleft - 1] = 0;
   233 	    return (bstr){outbuf, osize - oleft - 1};
   234 	#endif
   235 	
Target 0: (mpv) stopped.
(lldb) frame variable
(mp_log *) log = 0x00000002dc1affc0
(bstr) buf = (start = "", len = 0)
(const char *) cp = 0x00000001013adc0b "UTF-8-MAC"
(int) flags = 8
(iconv_t) icdsc = 0x00000002dc1cff70
(size_t) size = 0
(size_t) osize = 0
(size_t) ileft = 0
(size_t) oleft = 18446744073709551615
(char *) outbuf = 0x00000002dc1d8000 ""
(char *) ip = 0x0000000159843ff9 ""
(char *) op = 0x00000002dc1d8000 ""
(lldb) thread backtrace
* thread #18, name = 'mpv/worker', stop reason = EXC_BAD_ACCESS (code=1, address=0x2dc1d8000)
  * frame #0: 0x00000001000ecdec mpv`mp_iconv_to_utf8(log=0x00000002dc1affc0, buf=(start = "", len = 0), cp="UTF-8-MAC", flags=8) at charset_conv.c:232:31
    frame #1: 0x0000000100134bdc mpv`append_dir_subtitles(global=0x0000000119c5bfd0, opts=0x000000011a1b3ad0, slist=0x0000000286592e40, nsub=0x0000000286592e3c, path=(start = "https://www.tv2.no/video/nyhetene/satt-ut-av-oppvaskmaskin-triks/1810348/", len = 73), fname="https://www.tv2.no/video/nyhetene/satt-ut-av-oppvaskmaskin-triks/1810348/", limit_fuzziness=0, limit_type=-1) at external_files.c:189:27
    frame #2: 0x0000000100134990 mpv`find_external_files(global=0x0000000119c5bfd0, fname="https://www.tv2.no/video/nyhetene/satt-ut-av-oppvaskmaskin-triks/1810348/", opts=0x000000011a1b3ad0) at external_files.c:370:5
    frame #3: 0x00000001001393cc mpv`autoload_external_files(mpctx=0x0000000119c2bb40, cancel=0x0000000119c3bf40) at loadfile.c:894:26
    frame #4: 0x000000010013de14 mpv`load_external_opts_thread(p=0x0000000119c06c58) at loadfile.c:1399:5
    frame #5: 0x00000001000f1924 mpv`worker_thread(arg=0x0000000119c43f60) at thread_pool.c:88:9
    frame #6: 0x0000000188fd7fa8 libsystem_pthread.dylib`_pthread_start + 148
(lldb) frame select 1
frame #1: 0x0000000100134bdc mpv`append_dir_subtitles(global=0x0000000119c5bfd0, opts=0x000000011a1b3ad0, slist=0x0000000286592e40, nsub=0x0000000286592e3c, path=(start = "https://www.tv2.no/video/nyhetene/satt-ut-av-oppvaskmaskin-triks/1810348/", len = 73), fname="https://www.tv2.no/video/nyhetene/satt-ut-av-oppvaskmaskin-triks/1810348/", limit_fuzziness=0, limit_type=-1) at external_files.c:189:27
   186 	    struct mp_log *log = mp_log_new(tmpmem, global->log, "find_files");
   187 	
   188 	    struct bstr f_fbname = bstr0(mp_basename(fname));
-> 189 	    struct bstr f_fname = mp_iconv_to_utf8(log, f_fbname,
   190 	                                           "UTF-8-MAC", MP_NO_LATIN1_FALLBACK);
   191 	    struct bstr f_fname_noext = bstrdup(tmpmem, bstr_strip_ext(f_fname));
   192 	    bstr_lower(f_fname_noext);
(lldb) frame variable
(mpv_global *) global = 0x0000000119c5bfd0
(MPOpts *) opts = 0x000000011a1b3ad0
(subfn **) slist = 0x0000000286592e40
(int *) nsub = 0x0000000286592e3c
(bstr) path = (start = "https://www.tv2.no/video/nyhetene/satt-ut-av-oppvaskmaskin-triks/1810348/", len = 73)
(const char *) fname = 0x0000000159843fb0 "https://www.tv2.no/video/nyhetene/satt-ut-av-oppvaskmaskin-triks/1810348/"
(int) limit_fuzziness = 0
(int) limit_type = -1
(void *) tmpmem = 0x00000002dc1a8000
(mp_log *) log = 0x00000002dc1affc0
(bstr) f_fbname = (start = "", len = 0)
(bstr) f_fname = (start = "", len = 12282626016)
(bstr) f_fname_noext = (start = "\U00000018", len = 10843925808)
(bstr) f_fname_trim = (start = 0x0000000000000000, len = 12282626016)
(char *) path0 = 0x00000002dc19ff90 "\U00000018"
(DIR *) d = 0x0000000286592d00
(dirent *) de = NULL
(lldb) 
```
</details>

### Expected behavior
`mpv` can be run successfully using Guard Malloc.

### Actual behavior
EXC_BAD_ACCESS

### Log file
[mpv.log](https://github.com/mpv-player/mpv/files/11315703/mpv.log)

### Sample files
https://www.tv2.no/video/nyhetene/satt-ut-av-oppvaskmaskin-triks/1810348/

**Comment by forthrin** [comment]
Wow! Well spotted. If this is indeed the true culprit, it's been here for almost 10 years. It's stuff like this that explains the popularity of Rust. (Side note: I like the idea, but not the syntax.) I'll add the pull request pointer test before the culprit line with a printf to stderr and see if that concurs with the exception, though I assume the exception doesn't necessarily trigger every time because of the unpredictable order of memory allocations. Or are you able to trigger it every time?

```
commit f735a03346e8ec743bc89d5bdbaafd62dc0f084d
Author: wm4 <wm4@nowhere>
Date:   Sun Jun 23 22:15:04 2013 +0200

    sub: add subtitle charset conversion
    
```

**Comment by low-batt** [comment]
The fix proposed in PR #11627 is under active review.

I just streamed the example URL 3 times in a row using the released version of mpv. Played fine. As usual with heap corruption it depends upon what memory gets stepped on as to how and if the memory overwrite will manifest itself. The pattern of heap usage for applications that respond to asynchronous events and use multiple threads will vary from run to run. This makes memory corruption a critical problem that must be resolved.

The good news is that the developer tools shown above are usually effective in identifying the offending code. In the case at hand the problem was detected in every test run I did.

I took note of issue #11378 when it was first posted. I was busy at the time, so I only performed some limited testing which did not reproduce the problem. Since I had nothing to contribute I did not comment. A few days ago I was scanning open mpv issues looking to see if any new serious defects posted. I happened to check #11378 again and noticed the "keeps happening quite frequently" post. If something can be reproduced it can be tracked down. Thank you for being persistent and making sure this problem was noticed.
